### ISSUE TITLE
Bad joins performance 

### ISSUE THREAD

**Issue by melfa** [typeorm/typeorm#1952]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**
[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `websql`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

Performance of createQueryBuilder => innerJoinAndSelect => getMany two times less than createQueryBuilder => getMany. Without join performance is 2740 requests per second on Amazon EC2 c5.large instance (express+typeorm), and 1590 requests per second with join.

Performance test reports are here https://github.com/melfa/node-framework-load-test/tree/node8
Example source code https://github.com/melfa/node-framework-load-test/blob/node8/express-typeorm/src/index.ts

Tested on node 6.9/8.11, PostgreSQL 9.5/9.6/10 (but may affect all databases).

I do nodejs profiling and found the problem is regexp `/^[a-zA-Z0-9_-]+$/ which called numerous times for every resulting row. It is QueryBuilderUtils.isAliasProperty method. I do quick research and add caching to this method, and performance gap was fixed. Fixed code:
```typescript
export class QueryBuilderUtils {
    static readonly cache = new Map();

    /**
     * Checks if given value is a string representation of alias property,
     * e.g. "post.category" or "post.id".
     */
    static isAliasProperty(str: any): str is string {
        if (QueryBuilderUtils.cache.has(str)) {
            return QueryBuilderUtils.cache.get(str);
        }

        // alias property must be a string and must have a dot separator
        if (typeof str !== "string" || str.indexOf(".") === -1) {
            QueryBuilderUtils.cache.set(str, false);
            return false;
        }

        // extra alias and its property relation
        const [aliasName, propertyName] = str.split("."); // todo: what about relations in embedded?
        if (!aliasName || !propertyName) {
            QueryBuilderUtils.cache.set(str, false);
            return false;
        }

        // alias and property must be represented in a special format
        const aliasNameRegexp = /^[a-zA-Z0-9_-]+$/;
        if (!aliasNameRegexp.test(aliasName) || !aliasNameRegexp.test(propertyName)) {
            QueryBuilderUtils.cache.set(str, false);
            return false;
        }

        QueryBuilderUtils.cache.set(str, true);
        return true;
    }

}
```
Sure this solution is not production-ready, so i do not fork and pull request, but this is quick demonstration. If someone has suggestions how to fix this issue in right way i can spend few days and write better solution.

**Comment by pleerock** [comment]
Not sure why we need this check, I'll try to remove this regexp check and see what happens

**Comment by pleerock** [comment]
Fixed in `0.1.21`
